### PR TITLE
Fix requirement nesting

### PR DESCRIPTION
**Pull request by mcovarr**
Whilst working on env vars I noticed the Run / WorkflowStep / Workflow relationship wasn't modeled correctly wrt Requirements. That's something the env var conformance tests in particular will check, but it's generally useful to have this be correct so here it is in its own wee PR.

### CODE REVIEW

File: centaur/src/main/resources/standardTestCases/three_step.test [hunk 1]
```diff
@@ -16,5 +16,5 @@ metadata {
   "calls.cgrep.executionStatus": Done
   "calls.ps.runtimeAttributes.docker": "ubuntu:bionic"
   "calls.wc.runtimeAttributes.docker": "ubuntu:latest"
-  "calls.cgrep.runtimeAttributes.docker": "ubuntu:latest"
+  "calls.cgrep.runtimeAttributes.docker": "debian:jessie"
```
Comment (geoffjentry) [inline centaur/src/main/resources/standardTestCases/three_step.test]: pure curiosity here, why the change?

Comment (mcovarr) [reply centaur/src/main/resources/standardTestCases/three_step.test]: So I could assert that the `DockerRequirement` on the workflow step was correctly being given precedence over the `DockerRequirement` on the workflow.

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 2]
```diff
@@ -49,25 +49,27 @@ case class CommandLineTool private(
   private [cwl] implicit val explicitWorkflowName = ParentName(id)
   private val inputNames = this.inputs.map(i => FullyQualifiedName(i.id).id).toSet
 
+  private[cwl] var parentWorkflowStep: Option[WorkflowStep] = None
```
Comment (kshakir) [inline cwl/src/main/scala/cwl/CommandLineTool.scala]: _Shame!_ 🔔 🔔 

No really-- can you leave a comment beside this var with the briefest of descriptions for future developers as to why the `var` is required?

Comment (mcovarr) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Yeah not my proudest moment but I did what I had to. 🙂 I'll add some docs.

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 3]
```diff
@@ -49,25 +49,30 @@ case class CommandLineTool private(
   private [cwl] implicit val explicitWorkflowName = ParentName(id)
   private val inputNames = this.inputs.map(i => FullyQualifiedName(i.id).id).toSet
 
+  // Circe can't create bidirectional links between workflow steps and runs (including `CommandLineTool`s) so this
+  // ugly var is here to link back to a possible parent workflow step. This is needed to navigate upward for finding
+  // requirements in the containment hierarchy. There isn't always a containing workflow step so this is an `Option`.
+  private[cwl] var parentWorkflowStep: Option[WorkflowStep] = None
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CommandLineTool.scala]: Does this work even in a multiple file scenario? (I notice your test case is an all-in-one cwl file)

File: cwl/src/main/scala/cwl/Workflow.scala [hunk 4]
```diff
@@ -30,21 +30,24 @@ case class Workflow private(
...
+
+  steps.foreach { _.parentWorkflow = this }
+
   /** Builds an `Executable` from a `Workflow` CWL with no parent `Workflow` */
   def womExecutable(validator: RequirementsValidator, inputFile: Option[String] = None): Checked[Executable] = {
     CwlExecutableValidation.buildWomExecutable(womDefinition(validator), inputFile)
   }
 
-  private[cwl] var parentWorkflow: Option[Workflow] = None
+  private[cwl] var parentWorkflowStep: Option[WorkflowStep] = None
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/Workflow.scala]: ToL: This ugly var doesn't get the comment that the other two do

File: cwl/src/main/scala/cwl/CommandLineTool.scala (hunk 3 above)
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Doesn't SALAD all-in-one everything before we begin to parse?

File: cwl/src/main/scala/cwl/Workflow.scala (hunk 4 above)
Comment (mcovarr) [reply cwl/src/main/scala/cwl/Workflow.scala]: oops

**Comment by mcovarr** [comment]
@cjllanwarne while the ExpressionStep has requirement and hint fields it currently isn't doing anything with them at all. 🤔
